### PR TITLE
#124 reuse counter and max reuse

### PR DESCRIPTION
**Pull request by occupant23**

### CODE REVIEW

File: src/test/java/com/xceptance/neodymium/testclasses/webDriver/ValidateWebDriverMaxReuseWithTwoWebDriver.java [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+import com.xceptance.neodymium.NeodymiumRunner;
+import com.xceptance.neodymium.module.statement.browser.multibrowser.Browser;
+import com.xceptance.neodymium.module.statement.browser.multibrowser.WebDriverCache;
+import com.xceptance.neodymium.module.statement.browser.multibrowser.WebDriverStateContainer;
+import com.xceptance.neodymium.tests.NeodymiumTest;
+import com.xceptance.neodymium.tests.NeodymiumWebDriverTest;
+import com.xceptance.neodymium.util.Neodymium;
+
+@RunWith(NeodymiumRunner.class)
+public class ValidateWebDriverMaxReuseWithTwoWebDriver
```
Comment (ckeiner) [inline src/test/java/com/xceptance/neodymium/testclasses/webDriver/ValidateWebDriverMaxReuseWithTwoWebDriver.java:26]: - [ ] The purpose of this test class is not entirely clear. Maybe a comment would help.
- [ ] Typo. It's WebDriver**s**

File: src/main/java/com/xceptance/neodymium/module/statement/browser/multibrowser/WebDriverStateContainer.java [hunk 2]
```diff
@@ -33,4 +35,14 @@ public BrowserUpProxy getProxy()
     {
         return proxy;
     }
+
+    public int getUsedCount()
+    {
+        return usedCount;
+    }
+
+    public void increaseUsesedCount()
```
[the remark this answers is not in the document]
Comment (occupant23) [reply src/main/java/com/xceptance/neodymium/module/statement/browser/multibrowser/WebDriverStateContainer.java]: done

File: src/test/java/com/xceptance/neodymium/testclasses/webDriver/ValidateReuseWebDriver.java [hunk 3]
```diff
@@ -103,7 +104,9 @@ public void after()
     @AfterClass
     public static void afterClass()
     {
-        Assert.assertEquals(1, WebDriverCache.instance.getAllWebDriverAndProxy().size());
+        Assert.assertEquals(1, WebDriverCache.instance.getWebDriverStateContainerCacheSize());
+        WebDriverStateContainer cachingContainer = WebDriverCache.instance.getWebDriverStateContainerByBrowserTag("Chrome_headless");
```
[the remark this answers is not in the document]
Comment (occupant23) [reply src/test/java/com/xceptance/neodymium/testclasses/webDriver/ValidateReuseWebDriver.java:108]: changed it everywhere to `wDSContainer`.

File: src/main/java/com/xceptance/neodymium/module/statement/browser/BrowserStatement.java [hunk 4]
```diff
@@ -235,12 +232,24 @@ else if (Neodymium.configuration().reuseWebDriver() && !preventReuse && isWebDri
...
         Neodymium.setBrowserName(null);
     }
 
+    private boolean keepOpen(boolean testFailed, BrowserConfiguration browserConfiguration)
+    {
+        return (browserConfiguration != null && !browserConfiguration.isHeadless())
+               && ((Neodymium.configuration().keepBrowserOpenOnFailure() && testFailed) || Neodymium.configuration().keepBrowserOpen());
+    }
+
+    private boolean canReuse(boolean preventReuse, WebDriver webDriver, WebDriverStateContainer webDriverStateContainer)
```
[the remark this answers is not in the document]
Comment (occupant23) [reply src/main/java/com/xceptance/neodymium/module/statement/browser/BrowserStatement.java]: refactured it.

File: src/test/java/com/xceptance/neodymium/testclasses/webDriver/ValidateWebDriverMaxReuseWithTwoWebDriver.java (hunk 1 above)
Comment (occupant23) [reply src/test/java/com/xceptance/neodymium/testclasses/webDriver/ValidateWebDriverMaxReuseWithTwoWebDriver.java]: done

**Comment by occupant23** [comment]
> Your newly added test classes would greatly benefit from class comments since it is hard to follow what you actually want to test.

Added more documentation.

File: config/neodymium.properties [hunk 5]
```diff
@@ -196,6 +196,10 @@ neodymium.webDriver.firefox.legacyMode = false
 ## Whether to use a single driver instance for all tests (defaults to false).
 neodymium.webDriver.reuseDriver = false
 
+## The number of reuses. If not specified or set below 1 the driver will be reused unlimited times.
+## Setting the property to 1 means that the driver is reused once but twice in total.
```
Comment (ckeiner) [inline config/neodymium.properties:200]:
```suggestion
## Setting the property to 1 means that the driver is reused once, so used twice in total.
```
Otherwise it sounds like it is reused twice in total.

**Comment by oomelianchuk** [comment]
I think everything is fine. The test purposes are clear and the implementation is easy to understand.
